### ISSUE TITLE
[invalid] Batch training is flawed

### ISSUE THREAD

**Issue by sergevanhaag2** [tribhuvanesh/knockoffnets#4]
When training with different budgets in adversary/train.py, you do not reset the model each run. So if you want to train with a budget of 100 and then 200, you actually train with 100 and then 300 (100+200). Because the model has already seen the first 100 at the start of the second budget; the budget adds up. 

```
model = zoo.get_net(model_name, modelfamily, pretrained, num_classes=num_classes)
model = model.to(device)
for b in budgets:
    #train model
```

Solution: Reset or redefine the model each iteration.

**Comment by yxwang-10** [comment]
This is actually no problem.


```
> transferset = samples_to_transferset(transferset_samples, budget=b, transform=transform)
> def samples_to_transferset(samples, budget=None, transform=None, target_transform=None):
>     sample_x = samples[0][0]
>     assert budget <= len(samples), 'Required {} samples > Found {} samples'.format(budget, len(samples))
>     if isinstance(sample_x, str):
>         return TransferSetImagePaths(samples[:budget], transform=transform, target_transform=target_transform)
>     elif isinstance(sample_x, np.ndarray):
>         return TransferSetImages(samples[:budget], transform=transform, target_transform=target_transform)
>     else:
>         raise ValueError('type(x_i) ({}) not recognized. Supported types = (str, np.ndarray)'.format(type(sample_x)))
```


Each selected sample contains the previous sample``

**Comment by sergevanhaag2** [comment]
Yes, that's correct. However, it would regardless mess up the model, because with a budget of `--budgets 100, 200` and let's say `--epochs 10`, the model would train for 10 epochs on the first 100 samples and then for an additional 10 epochs on the 200 samples. In other words,  the model would have trained for 20 epochs on the first 100 samples, and then for 10 epochs on the next 100. It's just counterintuitive to me. 

Also, it would not be as much of a problem under the assumption that the batch size increases, however, it would not be the case if the batch size was specified as `--budgets 200, 5`. Then the model trained on a budget of `5` is actually a model trained on a budget of 200 (of which the first 5 samples would have been trained for 20 epochs, and the next 195 for 10 epochs).

**Comment by yxwang-10** [comment]
@tribhuvanesh You're right. This may be problematic. I tested these two strategies on the CIFAR10 data set, and the results without resetting would be better than resetting.

**Comment by sergevanhaag2** [comment]
Thank you for verifying, I appreciate it.
